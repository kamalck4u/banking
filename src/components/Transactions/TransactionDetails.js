import React from 'react';
import { HeaderPanel, GenericErrorPanel } from '../Common/Panels';
import Navigations from '../Common/Navigations'
import { REFERENCE_NUMBER, DATE, DESCRIPTION, STATUS, AMOUNT, TRANSACTIONS_DETAILS } from '../Common/Constants';

const TransactionDetails = (props) => {
	const transactionDetails = props.location.state ? props.location.state.transaction : undefined;
	return (
		<React.Fragment>
			<Navigations back />
			{
				transactionDetails ? <HeaderPanel header={TRANSACTIONS_DETAILS} content={transactionDetails.from} /> : <HeaderPanel />
			}
			{
				transactionDetails ?
					<div className="table-responsive">
						<table className="table table-striped">
							<thead className="thead-dark">
								<tr>
									<th scope="col">{REFERENCE_NUMBER}</th>
									<th scope="col">{DATE}</th>
									<th scope="col">{DESCRIPTION}</th>
									<th scope="col">{AMOUNT}</th>
									<th scope="col">{STATUS}</th>
								</tr>
							</thead>
							<tbody>
								<tr key={transactionDetails.id}>
									<td>{transactionDetails.id}</td>
									<td>{transactionDetails.transaction_date}</td>
									<td>{transactionDetails.description}</td>
									<td>{transactionDetails.amount}</td>
									<td>{transactionDetails.transaction_processed === true ? 'Processed' : 'Pending'}</td>
								</tr>
							</tbody>
						</table>
					</div>
					: <GenericErrorPanel />
			}
		</React.Fragment>
	);
};
export default TransactionDetails;