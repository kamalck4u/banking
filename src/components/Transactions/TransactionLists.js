import React from 'react';
import { useHistory } from 'react-router-dom';
import { REFERENCE_NUMBER, DATE, AMOUNT, LOADER_TEXT_TRANSACTIONS } from '../Common/Constants';
import { Loader } from '../Common/Panels';
import '../../App.css';
import { array, bool } from 'prop-types';

const TransactionLists = ({ transactions, isLoading }) => {
	const history = useHistory();
	return (
		<React.Fragment>
			{
				isLoading ? <Loader loaderText={LOADER_TEXT_TRANSACTIONS} /> :
					<div className="table-responsive">
						<table className="table table-striped">
							<thead className="thead-dark">
								<tr>
									<th scope="col">{REFERENCE_NUMBER}</th>
									<th scope="col">{DATE}</th>
									<th scope="col">{AMOUNT}</th>
								</tr>
							</thead>
							<tbody className="row-clickable">
								{transactions.map((transaction) => {
									return (
										<tr key={transaction.id} className="clickable-row table-cell" onClick={() => history.push(`/transactions/${transaction.id}/details`, { transaction })}>
											<td>{transaction.id}</td>
											<td>{transaction.transaction_date}</td>
											<td>{transaction.amount}</td>
										</tr>
									);
								}
								)}
							</tbody>
						</table>
					</div>
			}
		</React.Fragment>
	);
};

TransactionLists.propTypes = {
	transactions: array,
	isLoading: bool
};

export default TransactionLists;


