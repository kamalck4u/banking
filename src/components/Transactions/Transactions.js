import React, { useEffect, useState } from 'react';
import { FetchTransactions } from '../../apis/FetchAccountDetails';
import TransactionLists from './TransactionLists';
import { GenericErrorPanel, HeaderPanel } from '../Common/Panels';
import { TRANSACTIONS_HISTORY } from '../Common/Constants';

const Transactions = (props) => {
	const [transactions, setTransactions] = useState([]);
	const [error, setError] = useState({});
	const { id } = props.match.params;
	const account = props.location.state ? props.location.state.account : undefined;
	const [isLoading, setIsLoading] = useState();

	useEffect(() => {
		setIsLoading(true);
		(async () => {
			try {
				const response = await FetchTransactions(id);
				setTransactions(response.data);
			} catch (error) {
				setError({ isFailed: true, error: error });
			}
			setIsLoading(false);
		})();
	}, [id]);

	return (
		<React.Fragment>
			{
				account ? <HeaderPanel header={TRANSACTIONS_HISTORY} content={`${account.account_name} - ${account.account_number}`} /> : <HeaderPanel />
			}
			{
				error.isFailed
					? <GenericErrorPanel />
					: <TransactionLists transactions={transactions} isLoading={isLoading} />
			}
		</React.Fragment>
	);
};

export default Transactions;