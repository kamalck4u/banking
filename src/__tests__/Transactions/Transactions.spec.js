import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Transactions from '../../components/Transactions/Transactions';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const account = {
	'id': 1,
	'account_name': 'Bonus Savings Account',
	'account_type': 'savings',
	'balance': -5.11,
	'currency': 'AUD',
	'account_number': '156478245',
	'is_active': true
};


const match = {
	params: { id: 1 },
	isExact: true,
	path: '',
	url: ''
};

const location = {
	state: {
		account: account
	}
};

describe('Transactions', () => {
	it('snapshot renders', () => {
		const Transactionscomponent = renderer.create(<Transactions match={match} location={location} />);
		expect(Transactionscomponent.toJSON()).toMatchSnapshot();
	});
});

describe('Transactions Home page', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(<Transactions match={match} location={location} />);
	});

	it('should display account name & number on header panel', () => {
		expect(wrapper.find('HeaderPanel').props()).toEqual({ content: 'Bonus Savings Account - 156478245', header: 'Transaction History' });
	});

	it('should render AccountLists component', () => {
		expect(wrapper.find('TransactionLists')).toHaveLength(1);
	});

	it('should not render GenericErrorPanel component', () => {
		expect(wrapper.find('GenericErrorPanel')).toHaveLength(0);
	});
});

