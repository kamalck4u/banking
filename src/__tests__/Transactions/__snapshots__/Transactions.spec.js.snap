// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions snapshot renders 1`] = `
Array [
  <div
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <i
      aria-hidden="true"
      className="fa fa-home fa-2x nav-icon"
      onClick={[Function]}
    />
  </div>,
  <div
    className="jumbotron panel-header"
  >
    <h1
      className="display-4"
    >
      Transaction History
    </h1>
    <p
      className="lead"
    >
      Bonus Savings Account - 156478245
    </p>
    <hr
      className="my-4"
    />
    <p>
       
    </p>
  </div>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped"
    >
      <thead
        className="thead-dark"
      >
        <tr>
          <th
            scope="col"
          >
            Transaction#
          </th>
          <th
            scope="col"
          >
            Date
          </th>
          <th
            scope="col"
          >
            Amount
          </th>
        </tr>
      </thead>
      <tbody
        className="row-clickable"
      />
    </table>
  </div>,
]
`;
